---
title: "Small plasmid analysis"
date: "2021-02-22"
author: "Ryan Wick"
output:
  html_document:
    pandoc_args: ["+RTS", "-K64m", "-RTS", "--self-contained",]
    df_print: paged
    keep_md: false
    toc: true
    toc_float: true
---

This document contains the R code used to generate data and plots for my paper comparing rapid and ligation kits in the context of genomes with small plasmids.

It takes as input:

* `Table_S1.xlsx` - assumed to be in the same directory as this file.
* Four TSV files of read-level information, assumed to be in a `../data` directory:
  * `tech_rep_1_ligation_reads.tsv.gz`
  * `tech_rep_1_rapid_reads.tsv.gz`
  * `tech_rep_2_ligation_reads.tsv.gz`
  * `tech_rep_2_rapid_reads.tsv.gz`
* One TSV file of sliding-window GC and depth values: `gc_and_depth.tsv.gz`

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.




## Setup

```{r}
library(tidyverse)
library(knitr)
library(readxl)
library(cowplot)
library(ggrastr)
library(scales)
```

Output settings:
```{r}
opts_chunk$set(fig.width=8, fig.height=5, dpi=300, fig.path='../plots/', echo=TRUE, dev=c('png','pdf'), warning=FALSE, message=FALSE)
pdf.options(useDingbats = FALSE)
```




## Load data

Most of the data is pretty quick to load:
```{r}
barcode_data <- read_excel("Table_S1.xlsx", sheet = "Per-barcode", col_names = TRUE)
barcode_data$`Barcode number` <- factor(barcode_data$`Barcode number`, levels = barcode_data$`Barcode number`)

replicon_data <- read_excel("Table_S1.xlsx", sheet = "Per-replicon", col_names = TRUE)
replicon_data$size_category <- ifelse(replicon_data$Size >=20000, "big (â‰¥20 kbp)", "small (<20 kbp)")
replicon_data$rapid_to_illumina <- replicon_data$`ONT rapid read depth (normalised to chromosome)` / replicon_data$`Illumina read depth (normalised to chromosome)`
replicon_data$ligation_to_illumina <- replicon_data$`ONT ligation read depth (normalised to chromosome)` / replicon_data$`Illumina read depth (normalised to chromosome)`

run_data <- read_excel("Table_S1.xlsx", sheet = "Per-run", col_names = TRUE)
gc_depth <- read_delim("../data/gc_and_depth.tsv.gz", "\t", escape_double = FALSE, trim_ws = TRUE)
```

But the read data takes a while:
```{r}
load_read_data <- function(filename) {
  reads <- read_delim(filename, "\t", escape_double = FALSE, trim_ws = TRUE)
  reads$mean_identity <- as.numeric(sub("%", "", reads$mean_identity)) / 100.0
  reads$read_coverage <- as.numeric(sub("%", "", reads$read_coverage)) / 100.0
  reads$translocation_speed <- reads$read_length / reads$template_duration
  reads$start_time_hours <- reads$start_time / 3600
  return(reads)
}

tech_rep_1_ligation_reads <- load_read_data("../data/tech_rep_1_ligation_reads.tsv.gz")
tech_rep_1_rapid_reads <- load_read_data("../data/tech_rep_1_rapid_reads.tsv.gz")
tech_rep_2_ligation_reads <- load_read_data("../data/tech_rep_2_ligation_reads.tsv.gz")
tech_rep_2_rapid_reads <- load_read_data("../data/tech_rep_2_rapid_reads.tsv.gz")

```




## Yield

Here are the full yields:
```{r}
sum(tech_rep_1_ligation_reads$read_length, na.rm = TRUE)
sum(tech_rep_2_ligation_reads$read_length, na.rm = TRUE)
sum(tech_rep_1_rapid_reads$read_length, na.rm = TRUE)
sum(tech_rep_2_rapid_reads$read_length, na.rm = TRUE)
```

```{r}
max(tech_rep_1_ligation_reads$start_time_hours)
max(tech_rep_2_ligation_reads$start_time_hours)
max(tech_rep_1_rapid_reads$start_time_hours)
max(tech_rep_2_rapid_reads$start_time_hours)
```



Since not all runs went for the same amount of time, and some were refueled, which makes it harder to compare yields. So here are yields up to the 12-hour point (before any refueling took place):
```{r}
yield_before_hour <- function(read_table, hour_cutoff) {
  read_table %>% filter(start_time_hours < hour_cutoff) -> read_table
  return(sum(read_table$read_length))
}
```

```{r}
yield_before_hour(tech_rep_1_ligation_reads, 12)
yield_before_hour(tech_rep_2_ligation_reads, 12)
yield_before_hour(tech_rep_1_rapid_reads, 12)
yield_before_hour(tech_rep_2_rapid_reads, 12)
```





## Demultiplexing

```{r}
incorrect_demux_percent <- function(read_table) {
  
  # Filter out reads where the reference is the plasmid found in two genomes, as these are unreliable for demultiplexing.
  read_table = filter(read_table, reference_names != "Haemophilus_unknown_M1C132_1__plasmid_3")
  read_table = filter(read_table, reference_names != "Klebsiella_oxytoca_MSB1_2C__plasmid_3")

  correct <- sum(read_table$demultiplex_status == 'correct', na.rm = TRUE)
  incorrect <- sum(read_table$demultiplex_status == 'incorrect', na.rm = TRUE)
  
  incorrect_unused <- sum(read_table$demultiplex_status == 'incorrect' &
                          (read_table$barcode_arrangement == 'barcode06' |
                           read_table$barcode_arrangement == 'barcode09' |
                           read_table$barcode_arrangement == 'barcode10' |
                           read_table$barcode_arrangement == 'barcode11' |
                           read_table$barcode_arrangement == 'barcode12'), na.rm = TRUE)
  
  incorrect_used <- sum(read_table$demultiplex_status == 'incorrect' &
                        (read_table$barcode_arrangement == 'barcode01' |
                         read_table$barcode_arrangement == 'barcode02' |
                         read_table$barcode_arrangement == 'barcode03' |
                         read_table$barcode_arrangement == 'barcode04' |
                         read_table$barcode_arrangement == 'barcode05' |
                         read_table$barcode_arrangement == 'barcode07' |
                         read_table$barcode_arrangement == 'barcode08'), na.rm = TRUE)
  
  total <- correct + incorrect
  return(c(100.0 * incorrect_unused / total,
           100.0 * incorrect_used / total,
           100.0 * incorrect / total))
}
```

```{r}
incorrect_demux_percent(tech_rep_1_ligation_reads)
incorrect_demux_percent(tech_rep_2_ligation_reads)
incorrect_demux_percent(tech_rep_1_rapid_reads)
incorrect_demux_percent(tech_rep_2_rapid_reads)
```

I was curious is selecting for high quality reads fixed bad demuxing, so I tried creating high quality subsets based on `barcode_score`, `mean_qscore_template` and `mean_identity`. However, none of these reliably resulted in a significant improvement.





## Chimeras

```{r}
chimera_percent <- function(read_table) {
  total <- dim(read_table)[1]
  
  all_chimeras <- sum(read_table$chimera == 'yes', na.rm = TRUE)
  all_chimeras_percent <- 100.0 * all_chimeras / total
  
  within_bin_chimeras <- sum(read_table$within_bin_chimera == 'yes', na.rm = TRUE)
  within_bin_chimeras_percent <- 100.0 * within_bin_chimeras / total
  
  cross_bin_chimeras <- sum(read_table$cross_bin_chimera == 'yes', na.rm = TRUE)
  cross_bin_chimeras_percent <- 100.0 * cross_bin_chimeras / total
  
  return(c(within_bin_chimeras_percent,
           cross_bin_chimeras_percent,
           all_chimeras_percent))
}
```

```{r}
chimera_percent(tech_rep_1_ligation_reads)
chimera_percent(tech_rep_2_ligation_reads)
chimera_percent(tech_rep_1_rapid_reads)
chimera_percent(tech_rep_2_rapid_reads)
```




## ONT vs Illumina read depth

This plot puts plasmid size on the X axis and normalised depth ratios on the Y axis. The plot is done twice, once for ligation (left) and once for rapid (right).

```{r depth_vs_plasmid_size, fig.width = 8, fig.height = 3.75, useDingbats = FALSE}
depth_ratios <- data.frame(replicon_data$Replicon)
depth_ratios$tech_rep <- as.factor(replicon_data$`Technical replicate`)
depth_ratios$Size <- replicon_data$Size
depth_ratios$size_log <- log10(depth_ratios$Size)
depth_ratios$rapid_to_illumina <- replicon_data$rapid_to_illumina
depth_ratios$rapid_to_illumina_log <- log10(depth_ratios$rapid_to_illumina)
depth_ratios$ligation_to_illumina <- replicon_data$ligation_to_illumina
depth_ratios$ligation_to_illumina_log <- log10(depth_ratios$ligation_to_illumina)
colnames(depth_ratios) <- c("replicon", "tech_rep", "size", "size_log", "rapid", "rapid_log", "ligation", "ligation_log")
depth_ratios %>% filter(replicon != "chromosome") %>% drop_na() %>% select(-replicon) -> depth_ratios

size_depth_ratios <- function(depth_ratios, column_name, plot_title, y_title, point_colour, line_colour) {
  column_name <- sym(column_name)

  p <- ggplot(depth_ratios, aes(x = size, y = !!column_name)) +
    geom_hline(yintercept=1, linetype="dashed", colour = "#888888") +
    geom_point(aes(shape = tech_rep), colour = point_colour, size=2) +
    theme_bw() + theme(plot.background = element_blank(), legend.position = "none") +
    scale_x_log10(breaks = c(1000, 3000, 10000, 30000, 100000, 300000), labels = c("1", "3", "10", "30", "100", "300"), minor_breaks = NULL, expand=c(0, 0)) +
    scale_y_log10(breaks = c(0.01, 0.1, 1, 10), labels = c("0.01", "0.1", "1", "10"), minor_breaks = NULL) +
    coord_fixed(xlim = c(1000, 300000), ylim = c(0.007, 10), ratio=0.7) +
    labs(x = "Plasmid size (kbp)", y = y_title, title = plot_title)
    
  return(p)
}
  
p1 <- size_depth_ratios(depth_ratios, "ligation",
                        "Oxford Nanopore vs Illumina sequencing read depths for plasmids\n",
                        "Ligation:Illumina read depth ratio", "#CB4D42", "#760000")
p2 <- size_depth_ratios(depth_ratios, "rapid", "",
                        "Rapid:Illumina read depth ratio", "#00989D", "#00555A")

plot_grid(p1, p2, ncol = 2, align="hv")
```


## P-values

I wanted p-values for whether or not there is a relationship between plasmid size and the ONT/Illumina depth ratio. I.e. the null hypothesis is that plasmid size does not affect depth ratios.

One way to do it is a linear regression. I think this is best done on the log-transformed data, as that improves the assumptions of a least-squares regression.
```{r}
ligation_to_size <- lm(ligation_log ~ size_log, data=depth_ratios)
rapid_to_size <- lm(rapid_log ~ size_log, data=depth_ratios)

summary(ligation_to_size)
summary(rapid_to_size)
```
However, this regression implies a specific kind of relationship which I don't think exists. E.g. I would expect the relationship to level off with higher plasmid sizes.

So another way is a rank-based non-parametic correlation test, like Spearman/Kendall. I think this is better because it doesn't assume a shape of the relationship. It also gives the same answer whether or not I log the data (because that doesn't change the rank), which is nice. I went with Kendall because it can handle ties better (my data has ties).
```{r}
cor.test(depth_ratios$size, depth_ratios$ligation, method="kendall", alternative="two.sided")
cor.test(depth_ratios$size, depth_ratios$rapid, method="kendall", alternative="two.sided")
```

The results are pretty clear! For ligation (p = 5.76e-12), there is a strong relationship. For rapid (p = 0.474), there isn't.


Calculating the mean/min underrepresentation of different sized plasmids in ligation reads (I report this value in the text):
```{r}
large_plasmid_depth_ratios <- filter(depth_ratios, size > 20000)
mean(large_plasmid_depth_ratios$ligation)
min(large_plasmid_depth_ratios$ligation)
```


```{r}
small_plasmid_depth_ratios <- filter(depth_ratios, size < 20000)
mean(small_plasmid_depth_ratios$ligation)
min(small_plasmid_depth_ratios$ligation)
```

```{r}
very_small_plasmid_depth_ratios <- filter(depth_ratios, size < 3000)
mean(very_small_plasmid_depth_ratios$ligation)
min(very_small_plasmid_depth_ratios$ligation)
```


## Barcode distributions

```{r barcode_yields, fig.width = 8, fig.height = 8, useDingbats = FALSE}
barcode_distribution <- function(barcode_data, column_name, plot_title, fill_colour, max_y) {
  column_name <- sym(column_name)
  wrong_colour = "#999999"
  unclassified_colour = "#666666"
  
  p <- ggplot(barcode_data) +
    geom_bar(aes(x=`Barcode number`, y=!!column_name, fill=`Barcode number`), stat = "identity") +
    theme_bw() + theme(plot.background = element_blank()) +
    scale_fill_manual(values = c("1" = fill_colour, "2" = fill_colour, "3" = fill_colour,
                                 "4" = fill_colour, "5" = fill_colour, "6" = wrong_colour,
                                 "7" = fill_colour, "8" = fill_colour, "9" = wrong_colour,
                                 "10" = wrong_colour, "11" = wrong_colour, "12" = wrong_colour,
                                 "unclassified" = unclassified_colour)) +
    guides(fill = FALSE) +
    scale_y_log10(breaks = c(1, 10, 100, 1000, 10000, 100000, 1000000, 10000000),
                  labels = c(bquote(10^0), bquote(10^1), bquote(10^2), bquote(10^3), bquote(10^4), bquote(10^5), bquote(10^6), bquote(10^7)),
                  minor_breaks = NULL,
                  expand = c(0, 0)) +
    coord_cartesian(ylim = c(1, max_y)) +
    labs(x = "Barcode", y = "Read count", title = plot_title)
}


p1 <- barcode_distribution(barcode_data, "Ligation (run 1) read count", "Ligation (run 1)", "#CB4D42", 10000000)
p2 <- barcode_distribution(barcode_data, "Ligation (run 2) read count", "Ligation (run 2)", "#CB4D42", 10000000)
p3 <- barcode_distribution(barcode_data, "Rapid (run 1) read count", "Rapid (run 1)", "#00989D", 10000000)
p4 <- barcode_distribution(barcode_data, "Rapid (run 2) read count", "Rapid (run 2)", "#00989D", 10000000)

plot_grid(p1, p2, p3, p4, ncol = 1, align="hv")
```





## Barcode N50s

```{r barcode_n50s, fig.width = 8, fig.height = 5, useDingbats = FALSE}
barcode_n50s <- function(barcode_data, column_name, plot_title, fill_colour, max_y) {
  barcode_data_real_barcodes <- filter(barcode_data, `Barcode number` != "6" & `Barcode number` != "9" &
                                                     `Barcode number` != "10" & `Barcode number` != "11" &
                                                     `Barcode number` != "12" & `Barcode number` != "unclassified")

  column_name <- sym(column_name)
  wrong_colour = "#999999"
  unclassified_colour = "#666666"
  
  p <- ggplot(barcode_data_real_barcodes) +
    geom_bar(aes(x=`Barcode number`, y=!!column_name, fill=`Barcode number`), stat = "identity") +
    theme_bw() + theme(plot.background = element_blank()) +
    scale_fill_manual(values = c("1" = fill_colour, "2" = fill_colour, "3" = fill_colour,
                                 "4" = fill_colour, "5" = fill_colour, "6" = wrong_colour,
                                 "7" = fill_colour, "8" = fill_colour, "9" = wrong_colour,
                                 "10" = wrong_colour, "11" = wrong_colour, "12" = wrong_colour,
                                 "unclassified" = unclassified_colour)) +
    guides(fill = FALSE) +
    scale_y_continuous(expand = c(0, 0), breaks = c(0, 10000, 20000, 30000), labels = c("0", "10k", "20k", "30k")) + 
    coord_cartesian(ylim = c(0, max_y)) +
    labs(x = "Barcode", y = "Read N50", title = plot_title)
}


p1 <- barcode_n50s(barcode_data, "Ligation (run 1) read N50", "Ligation (run 1)", "#CB4D42", 30000)
p2 <- barcode_n50s(barcode_data, "Ligation (run 2) read N50", "Ligation (run 2)", "#CB4D42", 30000)
p3 <- barcode_n50s(barcode_data, "Rapid (run 1) read N50", "Rapid (run 1)", "#00989D", 30000)
p4 <- barcode_n50s(barcode_data, "Rapid (run 2) read N50", "Rapid (run 2)", "#00989D", 30000)

plot_grid(p1, p2, p3, p4, ncol = 2, align="hv")
```

This has the same data as the previous plot, but in a grouped bar format:
```{r barcode_n50s_grouped, fig.width = 8, fig.height = 4, useDingbats = FALSE}
barcode_data_real_barcodes <- filter(barcode_data, `Barcode number` != "6" & `Barcode number` != "9" &
                                                   `Barcode number` != "10" & `Barcode number` != "11" &
                                                   `Barcode number` != "12" & `Barcode number` != "unclassified")
barcode_data_real_barcodes = subset(barcode_data_real_barcodes, select = c(`Barcode number`,
                                                                           `Ligation (run 1) read N50`, `Rapid (run 1) read N50`,
                                                                           `Ligation (run 2) read N50`, `Rapid (run 2) read N50`))
colnames(barcode_data_real_barcodes) <- c("Num", "Ligation 1", "Rapid 1", "Ligation 2", "Rapid 2")
barcode_data_real_barcodes <- barcode_data_real_barcodes %>% pivot_longer(-Num, names_to = "Sequencing run", values_to = "N50")
barcode_data_real_barcodes$`Sequencing run` <- factor(barcode_data_real_barcodes$`Sequencing run`, levels = c("Ligation 1", "Rapid 1", "Ligation 2", "Rapid 2"))

ggplot(barcode_data_real_barcodes) +
    geom_bar(aes(x=Num, y=N50, fill=`Sequencing run`), stat = "identity", position = "dodge") +
    theme_bw() + theme(plot.background = element_blank()) +
    scale_fill_manual(values = c("Ligation 1" = "#CB4D42", "Rapid 1" = "#00989D",
                                 "Ligation 2" = "#E58E8C", "Rapid 2" = "#52CCCC")) +
    scale_y_continuous(expand = c(0, 0), breaks = c(0, 10000, 20000, 30000), labels = c("0", "10k", "20k", "30k")) +
    coord_cartesian(ylim = c(0, 30000)) +
    labs(x = "Barcode number", y = "Read N50", title = "Per-barcode read N50 lengths")
```




## Read identity distributions

```{r identity_distributions, fig.width = 8, fig.height = 8, useDingbats = FALSE}
identity_distribution <- function(read_table, plot_title, fill_colour, line_colour) {
  read_table %>% filter(mean_identity > 0.5) -> read_table
  read_table %>% filter(read_length > 10000) -> read_table
  read_table %>% filter(read_coverage > 0.9) -> read_table
  
  # Optionally, filter for reads with a good translocation speed.
  # read_table %>% filter(translocation_speed > 380) -> read_table
  # read_table %>% filter(translocation_speed < 400) -> read_table
  
  median_identity <- median(read_table$mean_identity)
  
  p <- ggplot(read_table) +
    geom_density(aes(x = mean_identity, y = ..scaled..), bw = 0.0025, fill = fill_colour, colour = NA) +
    geom_vline(aes(xintercept = median_identity), alpha = 0.75, linetype = "dashed", colour = line_colour) + 
    theme_bw() +  theme(plot.background = element_blank()) +
    coord_cartesian(xlim = c(0.85, 1.0), ylim = c(0, 1.1)) +
    scale_x_continuous(breaks = seq(0.85, 1, 0.01),
                       labels = paste(seq(85, 100, 1), "%", sep = ""),
                       expand = c(0.002, 0.002)) +
    scale_y_continuous(breaks = NULL, expand = c(0, 0)) +
    labs(title = plot_title, x = "read identity", y = NULL)
  
  return(p)
}

proportion_under_identity <- function(read_table, identity_threshold) {
  read_table %>% filter(mean_identity > 0.5) -> read_table
  read_table %>% filter(read_length > 10000) -> read_table
  read_table %>% filter(read_coverage > 0.9) -> read_table
  
  total_reads <- nrow(read_table)
  read_under_90 <- nrow(read_table %>% filter(mean_identity < identity_threshold))
  
  return(read_under_90 / total_reads)
}

p1 <- identity_distribution(tech_rep_1_ligation_reads, "Ligation (run 1)", "#F8766D", "#760000")
p2 <- identity_distribution(tech_rep_2_ligation_reads, "Ligation (run 2)", "#F8766D", "#760000")
p3 <- identity_distribution(tech_rep_1_rapid_reads, "Rapid (run 1)", "#00BFC4", "#00555A")
p4 <- identity_distribution(tech_rep_2_rapid_reads, "Rapid (run 2)", "#00BFC4", "#00555A")

plot_grid(p1, p2, p3, p4, ncol = 1, align="hv")

proportion_under_identity(tech_rep_1_ligation_reads, 0.9)
proportion_under_identity(tech_rep_2_ligation_reads, 0.9)
proportion_under_identity(tech_rep_1_rapid_reads, 0.9)
proportion_under_identity(tech_rep_2_rapid_reads, 0.9)
```




## Translocation speed distributions

```{r translocation_speed_distributions, fig.width = 8, fig.height = 8, useDingbats = FALSE}
translocation_speed_distribution <- function(read_table, plot_title, fill_colour) {
  read_table %>% filter(mean_identity > 0.5) -> read_table
  read_table %>% filter(read_length > 10000) -> read_table
  
  p <- ggplot(read_table, aes(translocation_speed)) +
    geom_density(fill = fill_colour, colour = NA) +
    theme_bw() + theme(plot.background = element_blank()) +
    coord_cartesian(xlim = c(0, 600)) +
    scale_x_continuous(breaks = seq(0, 600, 100),
                       expand = c(0.02, 0.02)) +
    scale_y_continuous(breaks = NULL) +
    labs(title = plot_title, x = NULL, y = NULL)
  
  return(p)
}

p1 <- translocation_speed_distribution(tech_rep_1_ligation_reads, "Ligation 1 translocation speed", "#F8766D")
p2 <- translocation_speed_distribution(tech_rep_2_ligation_reads, "Ligation 2 translocation speed", "#F8766D")
p3 <- translocation_speed_distribution(tech_rep_1_rapid_reads, "Rapid 3 translocation speed", "#00BFC4")
p4 <- translocation_speed_distribution(tech_rep_2_rapid_reads, "Rapid 4 translocation speed", "#00BFC4")

plot_grid(p1, p2, p3, p4, ncol = 1, align="hv")
```




## Identity vs translocation speed

```{r identity_vs_translocation_speed_plots, fig.width = 8, fig.height = 8, useDingbats = FALSE}
identity_vs_translocation_speed <- function(read_table, plot_title, point_colour) {
  read_table %>% filter(mean_identity > 0.5) -> read_table
  read_table %>% filter(read_length > 10000) -> read_table
  
  read_table <- read_table[0:50000,]  # so we don't have too many points

  p <- ggplot(read_table) +
    rasterise(geom_point(aes(x=translocation_speed, y=mean_identity), colour=point_colour, alpha=0.15, shape=16, size=0.5), dpi=600) +
    theme_bw() + theme(plot.background = element_blank()) +
    scale_y_continuous(breaks = seq(0.9, 1, 0.02),
                       labels = paste(seq(90, 100, 2), "%", sep = "")) +
    scale_x_continuous(breaks = seq(0, 600, 100)) +
    coord_cartesian(xlim = c(0, 600), ylim = c(0.9, 1)) +
    labs(title = plot_title, x = "translocation speed (bp/sec)", y = "identity")
  
  return(p)
}

p1 <- identity_vs_translocation_speed(tech_rep_1_ligation_reads, "Ligation (run 1)", "#CB4D42")
p2 <- identity_vs_translocation_speed(tech_rep_2_ligation_reads, "Ligation (run 2)", "#CB4D42")
p3 <- identity_vs_translocation_speed(tech_rep_1_rapid_reads, "Rapid (run 1)", "#00989D")
p4 <- identity_vs_translocation_speed(tech_rep_2_rapid_reads, "Rapid (run 2)", "#00989D")

plot_grid(p1, p2, p3, p4, ncol = 2, align="hv")
```




## Identity vs read time

```{r identity_vs_read_time_plots, fig.width = 8, fig.height = 8, useDingbats = FALSE}
max_time <- 24

identity_vs_start_time <- function(read_table, plot_title, point_colour) {
  read_table %>% filter(start_time_hours < max_time) -> read_table
  read_table %>% filter(mean_identity > 0.5) -> read_table
  read_table %>% filter(read_length > 10000) -> read_table
  
  read_table <- read_table[0:100000,]  # so we don't have too many points

  p <- ggplot(read_table) +
    rasterise(geom_point(aes(x=start_time_hours, y=mean_identity), colour=point_colour, alpha=0.05, shape=16, size=0.5), dpi=600) +
    theme_bw() + theme(plot.background = element_blank()) +
    scale_y_continuous(breaks = seq(0.9, 1, 0.02),
                       labels = paste(seq(90, 100, 2), "%", sep = "")) +
    scale_x_continuous(breaks = seq(0, max_time, 6),
                       expand = c(0, 0)) +
    coord_cartesian(xlim = c(0, max_time), ylim = c(0.92, 1)) +
    labs(title = plot_title, x = "read time (hours)", y = "identity")
  
  return(p)
}

p1 <- identity_vs_start_time(tech_rep_1_ligation_reads, "Ligation (run 1)", "#CB4D42")
p2 <- identity_vs_start_time(tech_rep_2_ligation_reads, "Ligation (run 2)", "#CB4D42")
p3 <- identity_vs_start_time(tech_rep_1_rapid_reads, "Rapid (run 1)", "#00989D")
p4 <- identity_vs_start_time(tech_rep_2_rapid_reads, "Rapid (run 2)", "#00989D")

plot_grid(p1, p2, p3, p4, ncol = 2, align="hv")
```




## Translocation speed vs read time

Trying the same idea as the previous one, but now with the read's start time.

```{r translocation_speed_vs_read_time_plots, fig.width = 8, fig.height = 8, useDingbats = FALSE}
max_time <- 24

translocation_speed_vs_start_time <- function(read_table, plot_title, point_colour) {
  read_table %>% filter(start_time_hours < max_time) -> read_table
  read_table %>% filter(mean_identity > 0.5) -> read_table
  read_table %>% filter(read_length > 10000) -> read_table
  
  read_table <- read_table[0:50000,]  # so we don't have too many points

  p <- ggplot(read_table) +
    rasterise(geom_point(aes(x=start_time_hours, y=translocation_speed), colour=point_colour, alpha=0.15, shape=16, size=0.25), dpi=600) +
    theme_bw() + theme(plot.background = element_blank()) +
    scale_y_continuous(breaks = seq(0, 600, 100)) +
    scale_x_continuous(breaks = seq(0, max_time, 6),
                       expand = c(0, 0)) +
    coord_cartesian(xlim = c(0, max_time), ylim = c(0, 600)) +
    labs(title = plot_title, x = "read time (hours)", y = "translocation speed (bp/sec)")
  
  return(p)
}

p1 <- translocation_speed_vs_start_time(tech_rep_1_ligation_reads, "Ligation (run 1)", "#CB4D42")
p2 <- translocation_speed_vs_start_time(tech_rep_2_ligation_reads, "Ligation (run 2)", "#CB4D42")
p3 <- translocation_speed_vs_start_time(tech_rep_1_rapid_reads, "Rapid (run 1)", "#00989D")
p4 <- translocation_speed_vs_start_time(tech_rep_2_rapid_reads, "Rapid (run 2)", "#00989D")

plot_grid(p1, p2, p3, p4, ncol = 2, align="hv")
```




## Illumina GC bias

```{r illumina_gc_plot, fig.width = 7, fig.height = 5, useDingbats = FALSE}
replicon_data %>% filter(Replicon != "chromosome") -> plasmid_data
plasmid_gc_values <- data.frame(sort(unique(plasmid_data$GC)) * 100)
colnames(plasmid_gc_values) <- c("GC")

p1 <- ggplot(plasmid_gc_values, aes(x = GC)) +
  geom_dotplot(binwidth = 0.5, fill = "#00B0F6", colour = "#0054AA") +
  theme_bw() +
  theme(plot.background = element_blank(),
        axis.text.x=element_blank(), axis.ticks.x=element_blank(), panel.border = element_blank(),
        axis.title.y = element_text(colour = "#0054AA"),
        plot.margin = margin(0, 0, 0, 0)) +
  labs(title = "Illumina GC bias", x = NULL, y = "plasmids") +
  scale_x_continuous(breaks = NULL, expand = c(0, 0)) +
  scale_y_continuous(breaks = NULL) +
  coord_cartesian(xlim = c(28.2, 71.8))

p2 <- ggplot(gc_depth, aes(x=gc, y=depth)) +
  rasterise(geom_point(colour="#6FA100", alpha=0.2, shape=16, size=0.3), dpi=600) +
  geom_smooth(method="loess", colour="#235400", size=0.75, se = FALSE, fullrange = TRUE) +
  theme_bw() +
  theme(plot.background = element_blank(),
        plot.margin = margin(0, 0, 0, 0)) +
  scale_x_continuous(breaks = seq(0, 100, 5), expand = c(0, 0)) +
  scale_y_continuous(breaks = seq(0, 2, 0.1), expand = c(0, 0)) +
  coord_cartesian(xlim = c(28.2, 71.8), ylim = c(0.5, 1.5)) +
  labs(x = "GC content (%)", y = "Relative read depth")

plot_grid(p1, p2, ncol = 1, rel_heights = c(1, 4), align="hv")
```




## Read-start count distributions

```{r read_start_distributions_plot, fig.width = 23, fig.height = 35, useDingbats = FALSE}
rapid_reads_per_window_distributions <- read_delim("../data/rapid_reads_per_window_distributions.tsv", "\t", escape_double = FALSE, trim_ws = TRUE)
ligation_reads_per_window_distributions <- read_delim("../data/ligation_reads_per_window_distributions.tsv", "\t", escape_double = FALSE, trim_ws = TRUE)

mysqrt_trans <- function() {
    trans_new("mysqrt", 
              transform = base::sqrt,
              inverse = function(x) ifelse(x<0, 0, x^2),
              domain = c(0, Inf))
}

plot_distributions <- function(table, row, colour, title) {
  replicon <- table[[row, "replicon"]]
  mean_reads <- table[[row, "mean_reads_per_window"]]
  empirical_counts <- as.numeric(unlist(str_split(table[[row, "reads_per_window_distribution"]], ",")))
  total <- sum(empirical_counts)
  empirical <- empirical_counts / total
  max_count <- length(empirical) - 1
  read_counts <- 0:max_count
  poisson <- dpois(read_counts, mean_reads)
  max_prob <- max(empirical, poisson)
  distributions <- data.frame(read_counts, empirical, poisson)
  distributions %>% pivot_longer(!read_counts, names_to = "type", values_to = "mass") -> distributions
  
  # The probability of seeing a read count as high as (or higher than) the maximum observed count.
  prob_of_max <- ppois(max_count-1, mean_reads, lower.tail = FALSE)
  prob_of_max_str <- formatC(prob_of_max, format = "e", digits = 2)
  prob_of_max_str <- paste("italic(P) (italic(X) >=", max_count, ")==", prob_of_max_str)

  p <- ggplot(distributions, aes(x=read_counts, y=mass, colour=type)) +
    # geom_vline(xintercept = max_count, colour="grey", linetype = "longdash") +
    geom_point() +
    annotate("text", label = prob_of_max_str, x = Inf, y = Inf, vjust = 1.3, hjust = 1.03, parse=TRUE) + 
    theme_bw() +
    scale_colour_manual(values = c("empirical" = colour, "poisson" = "#aaaaaa")) +
    theme(plot.background = element_blank(), plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"), legend.position = "none") +
    labs(title = title, x = NULL, y = NULL) +
    scale_x_continuous(trans="mysqrt", limits=c(0, NA)) +
    scale_y_continuous(trans="mysqrt", limits=c(0, max_prob*1.25))
  
  return(p)
}

lig_ab1 <- plot_distributions(ligation_reads_per_window_distributions,  1, "#CB4D42", expression(paste(italic("A. baumannii"), " J9 chromosome")))
lig_ab2 <- plot_distributions(ligation_reads_per_window_distributions,  2, "#CB4D42", expression(paste(italic("A. baumannii"), " J9 plasmid 1")))
lig_ab3 <- plot_distributions(ligation_reads_per_window_distributions,  3, "#CB4D42", expression(paste(italic("A. baumannii"), " J9 plasmid 2")))
lig_ck1 <- plot_distributions(ligation_reads_per_window_distributions,  4, "#CB4D42", expression(paste(italic("C. koseri"), " MINF_9D chromosome")))
lig_ck2 <- plot_distributions(ligation_reads_per_window_distributions,  5, "#CB4D42", expression(paste(italic("C. koseri"), " MINF_9D plasmid 1")))
lig_ck3 <- plot_distributions(ligation_reads_per_window_distributions,  6, "#CB4D42", expression(paste(italic("C. koseri"), " MINF_9D plasmid 2")))
lig_ek1 <- plot_distributions(ligation_reads_per_window_distributions,  7, "#CB4D42", expression(paste(italic("E. kobei"), " MSB1_1B chromosome")))
lig_ek2 <- plot_distributions(ligation_reads_per_window_distributions,  8, "#CB4D42", expression(paste(italic("E. kobei"), " MSB1_1B plasmid 1")))
lig_ek3 <- plot_distributions(ligation_reads_per_window_distributions,  9, "#CB4D42", expression(paste(italic("E. kobei"), " MSB1_1B plasmid 2")))
lig_ek4 <- plot_distributions(ligation_reads_per_window_distributions, 10, "#CB4D42", expression(paste(italic("E. kobei"), " MSB1_1B plasmid 3")))
lig_ek5 <- plot_distributions(ligation_reads_per_window_distributions, 11, "#CB4D42", expression(paste(italic("E. kobei"), " MSB1_1B plasmid 4")))
lig_hu1 <- plot_distributions(ligation_reads_per_window_distributions, 14, "#CB4D42", expression(paste(italic("Haemophilus"), " M1C132_1 chromosome")))
lig_hu2 <- plot_distributions(ligation_reads_per_window_distributions, 15, "#CB4D42", expression(paste(italic("Haemophilus"), " M1C132_1 plasmid 1")))
lig_hu3 <- plot_distributions(ligation_reads_per_window_distributions, 16, "#CB4D42", expression(paste(italic("Haemophilus"), " M1C132_1 plasmid 2")))
lig_hu4 <- plot_distributions(ligation_reads_per_window_distributions, 17, "#CB4D42", expression(paste(italic("Haemophilus"), " M1C132_1 plasmid 3")))
lig_ko1 <- plot_distributions(ligation_reads_per_window_distributions, 20, "#CB4D42", expression(paste(italic("K. oxytoca"), " MSB1_2C chromosome")))
lig_ko2 <- plot_distributions(ligation_reads_per_window_distributions, 21, "#CB4D42", expression(paste(italic("K. oxytoca"), " MSB1_2C plasmid 1")))
lig_ko3 <- plot_distributions(ligation_reads_per_window_distributions, 22, "#CB4D42", expression(paste(italic("K. oxytoca"), " MSB1_2C plasmid 2")))
lig_ko4 <- plot_distributions(ligation_reads_per_window_distributions, 23, "#CB4D42", expression(paste(italic("K. oxytoca"), " MSB1_2C plasmid 3")))
lig_kv1 <- plot_distributions(ligation_reads_per_window_distributions, 24, "#CB4D42", expression(paste(italic("K. variicola"), " INF345 chromosome")))
lig_kv2 <- plot_distributions(ligation_reads_per_window_distributions, 25, "#CB4D42", expression(paste(italic("K. variicola"), " INF345 plasmid 1")))
lig_kv3 <- plot_distributions(ligation_reads_per_window_distributions, 26, "#CB4D42", expression(paste(italic("K. variicola"), " INF345 plasmid 2")))
lig_kv4 <- plot_distributions(ligation_reads_per_window_distributions, 27, "#CB4D42", expression(paste(italic("K. variicola"), " INF345 plasmid 3")))
lig_kv5 <- plot_distributions(ligation_reads_per_window_distributions, 28, "#CB4D42", expression(paste(italic("K. variicola"), " INF345 plasmid 4")))
lig_kv6 <- plot_distributions(ligation_reads_per_window_distributions, 29, "#CB4D42", expression(paste(italic("K. variicola"), " INF345 plasmid 5")))
lig_sm1 <- plot_distributions(ligation_reads_per_window_distributions, 30, "#CB4D42", expression(paste(italic("S. marcescens"), " 17-147-1671 chromosome")))
lig_sm2 <- plot_distributions(ligation_reads_per_window_distributions, 31, "#CB4D42", expression(paste(italic("S. marcescens"), " 17-147-1671 plasmid 1")))
lig_sm3 <- plot_distributions(ligation_reads_per_window_distributions, 32, "#CB4D42", expression(paste(italic("S. marcescens"), " 17-147-1671 plasmid 2")))
lig_sm4 <- plot_distributions(ligation_reads_per_window_distributions, 35, "#CB4D42", expression(paste(italic("S. marcescens"), " 17-147-1671 plasmid 4")))

rap_ab1 <- plot_distributions(rapid_reads_per_window_distributions,  1, "#00989D", expression(paste(italic("A. baumannii"), " J9 chromosome")))
rap_ab2 <- plot_distributions(rapid_reads_per_window_distributions,  2, "#00989D", expression(paste(italic("A. baumannii"), " J9 plasmid 1")))
rap_ab3 <- plot_distributions(rapid_reads_per_window_distributions,  3, "#00989D", expression(paste(italic("A. baumannii"), " J9 plasmid 2")))
rap_ck1 <- plot_distributions(rapid_reads_per_window_distributions,  4, "#00989D", expression(paste(italic("C. koseri"), " MINF_9D chromosome")))
rap_ck2 <- plot_distributions(rapid_reads_per_window_distributions,  5, "#00989D", expression(paste(italic("C. koseri"), " MINF_9D plasmid 1")))
rap_ck3 <- plot_distributions(rapid_reads_per_window_distributions,  6, "#00989D", expression(paste(italic("C. koseri"), " MINF_9D plasmid 2")))
rap_ek1 <- plot_distributions(rapid_reads_per_window_distributions,  7, "#00989D", expression(paste(italic("E. kobei"), " MSB1_1B chromosome")))
rap_ek2 <- plot_distributions(rapid_reads_per_window_distributions,  8, "#00989D", expression(paste(italic("E. kobei"), " MSB1_1B plasmid 1")))
rap_ek3 <- plot_distributions(rapid_reads_per_window_distributions,  9, "#00989D", expression(paste(italic("E. kobei"), " MSB1_1B plasmid 2")))
rap_ek4 <- plot_distributions(rapid_reads_per_window_distributions, 10, "#00989D", expression(paste(italic("E. kobei"), " MSB1_1B plasmid 3")))
rap_ek5 <- plot_distributions(rapid_reads_per_window_distributions, 11, "#00989D", expression(paste(italic("E. kobei"), " MSB1_1B plasmid 4")))
rap_hu1 <- plot_distributions(rapid_reads_per_window_distributions, 14, "#00989D", expression(paste(italic("Haemophilus"), " M1C132_1 chromosome")))
rap_hu2 <- plot_distributions(rapid_reads_per_window_distributions, 15, "#00989D", expression(paste(italic("Haemophilus"), " M1C132_1 plasmid 1")))
rap_hu3 <- plot_distributions(rapid_reads_per_window_distributions, 16, "#00989D", expression(paste(italic("Haemophilus"), " M1C132_1 plasmid 2")))
rap_hu4 <- plot_distributions(rapid_reads_per_window_distributions, 17, "#00989D", expression(paste(italic("Haemophilus"), " M1C132_1 plasmid 3")))
rap_hu5 <- plot_distributions(rapid_reads_per_window_distributions, 18, "#00989D", expression(paste(italic("Haemophilus"), " M1C132_1 plasmid 4")))
rap_hu6 <- plot_distributions(rapid_reads_per_window_distributions, 19, "#00989D", expression(paste(italic("Haemophilus"), " M1C132_1 plasmid 5")))
rap_ko1 <- plot_distributions(rapid_reads_per_window_distributions, 20, "#00989D", expression(paste(italic("K. oxytoca"), " MSB1_2C chromosome")))
rap_ko2 <- plot_distributions(rapid_reads_per_window_distributions, 21, "#00989D", expression(paste(italic("K. oxytoca"), " MSB1_2C plasmid 1")))
rap_ko3 <- plot_distributions(rapid_reads_per_window_distributions, 22, "#00989D", expression(paste(italic("K. oxytoca"), " MSB1_2C plasmid 2")))
rap_ko4 <- plot_distributions(rapid_reads_per_window_distributions, 23, "#00989D", expression(paste(italic("K. oxytoca"), " MSB1_2C plasmid 3")))
rap_kv1 <- plot_distributions(rapid_reads_per_window_distributions, 24, "#00989D", expression(paste(italic("K. variicola"), " INF345 chromosome")))
rap_kv2 <- plot_distributions(rapid_reads_per_window_distributions, 25, "#00989D", expression(paste(italic("K. variicola"), " INF345 plasmid 1")))
rap_kv3 <- plot_distributions(rapid_reads_per_window_distributions, 26, "#00989D", expression(paste(italic("K. variicola"), " INF345 plasmid 2")))
rap_kv4 <- plot_distributions(rapid_reads_per_window_distributions, 27, "#00989D", expression(paste(italic("K. variicola"), " INF345 plasmid 3")))
rap_kv5 <- plot_distributions(rapid_reads_per_window_distributions, 28, "#00989D", expression(paste(italic("K. variicola"), " INF345 plasmid 4")))
rap_kv6 <- plot_distributions(rapid_reads_per_window_distributions, 29, "#00989D", expression(paste(italic("K. variicola"), " INF345 plasmid 5")))
rap_sm1 <- plot_distributions(rapid_reads_per_window_distributions, 30, "#00989D", expression(paste(italic("S. marcescens"), " 17-147-1671 chromosome")))
rap_sm2 <- plot_distributions(rapid_reads_per_window_distributions, 31, "#00989D", expression(paste(italic("S. marcescens"), " 17-147-1671 plasmid 1")))
rap_sm3 <- plot_distributions(rapid_reads_per_window_distributions, 32, "#00989D", expression(paste(italic("S. marcescens"), " 17-147-1671 plasmid 2")))
rap_sm4 <- plot_distributions(rapid_reads_per_window_distributions, 35, "#00989D", expression(paste(italic("S. marcescens"), " 17-147-1671 plasmid 4")))

plot_grid(lig_ab1, lig_ab2, lig_ab3, NULL, NULL, NULL,
          lig_ck1, lig_ck2, lig_ck3, NULL, NULL, NULL,
          lig_ek1, lig_ek2, lig_ek3, lig_ek4, lig_ek5, NULL,
          lig_hu1, lig_hu2, lig_hu3, lig_hu4, NULL, NULL,
          lig_ko1, lig_ko2, lig_ko3, lig_ko4, NULL, NULL,
          lig_kv1, lig_kv2, lig_kv3, lig_kv4, lig_kv5, lig_kv6,
          lig_sm1, lig_sm2, lig_sm3, lig_sm4, NULL, NULL,
          rap_ab1, rap_ab2, rap_ab3, NULL, NULL, NULL,
          rap_ck1, rap_ck2, rap_ck3, NULL, NULL, NULL,
          rap_ek1, rap_ek2, rap_ek3, rap_ek4, rap_ek5, NULL,
          rap_hu1, rap_hu2, rap_hu3, rap_hu4, rap_hu5, rap_hu6,
          rap_ko1, rap_ko2, rap_ko3, rap_ko4, NULL, NULL,
          rap_kv1, rap_kv2, rap_kv3, rap_kv4, rap_kv5, rap_kv6,
          rap_sm1, rap_sm2, rap_sm3, rap_sm4, NULL, NULL,
          ncol = 6, align="hv")
```




## Read-start count Manhattan plots


```{r read_start_manhattan_plot, fig.width = 12, fig.height = 2.5, useDingbats = FALSE}
signed_sqrt <- function(x){sign(x)*sqrt(abs(x))}
inv_signed_sqrt <- function(x){x^2*sign(x)}
mysqrt_trans <- function() {
    trans_new("mysqrt",
              transform = signed_sqrt,
              inverse = inv_signed_sqrt)
}

manhattan_plot <- function(directory, replicon, colour, sig_threshold, title, replicon_size) {
  filename <- paste(directory, "/", replicon, ".tsv", sep = "")
  counts <- read_delim(filename, "\t", escape_double = FALSE, trim_ws = TRUE)
  counts %>% filter(signed_neg_log10_p_val != "n/a") -> counts
  counts$signed_neg_log10_p_val <- as.numeric(counts$signed_neg_log10_p_val)

  # Cap values at 100 (or -100).
  counts$signed_neg_log10_p_val[counts$signed_neg_log10_p_val > 100] <- 100
  counts$signed_neg_log10_p_val[counts$signed_neg_log10_p_val < -100] <- -100

  ratio <- 0.5 * (replicon_size^0.75)

  ggplot(counts, aes(x=window, y=signed_neg_log10_p_val)) +
      rasterise( geom_point(colour=colour), dpi=600) +
      geom_hline(yintercept=sig_threshold, linetype = "dashed") +
      geom_hline(yintercept=-sig_threshold, linetype = "dashed") +
      theme_bw() +
      theme(plot.background = element_blank(), plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"), legend.position = "none") +
      labs(title = title, x = NULL, y = NULL) +
      scale_x_continuous(labels = NULL, breaks = NULL, expand = c(0, 0), limits=c(0, replicon_size)) +
      scale_y_continuous(trans="mysqrt", limits=c(-100, 100), breaks = c(-100, -50, -10, 0, 10, 50, 100), minor_breaks = NULL,
                         labels = c("100", "50", "10","0", "10", "50", "100")) +
      coord_fixed(ratio = ratio)
}

manhattan_plot("../data/ligation_read_start_counts", "Acinetobacter_baumannii_J9__chromosome", "#CB4D42", 5.879233546813747, expression(paste(italic("A. baumannii"), " chromosome")), 3798646)
manhattan_plot("../data/ligation_read_start_counts", "Acinetobacter_baumannii_J9__plasmid_1", "#CB4D42", 4.418301291319746, expression(paste(italic("A. baumannii"), " plasmid 1")), 145059)
manhattan_plot("../data/ligation_read_start_counts", "Acinetobacter_baumannii_J9__plasmid_2", "#CB4D42", 3.0791812460476247, expression(paste(italic("A. baumannii"), " plasmid 2")), 6078)
manhattan_plot("../data/ligation_read_start_counts", "Citrobacter_koseri_MINF_9D__chromosome", "#CB4D42", 5.976881629799957, expression(paste(italic("C. koseri"), " chromosome")), 4758908)
manhattan_plot("../data/ligation_read_start_counts", "Citrobacter_koseri_MINF_9D__plasmid_1", "#CB4D42", 4.083502619830267, expression(paste(italic("C. koseri"), " plasmid 1")), 64962)
manhattan_plot("../data/ligation_read_start_counts", "Citrobacter_koseri_MINF_9D__plasmid_2", "#CB4D42", 3.2648178230095364, expression(paste(italic("C. koseri"), " plasmid 2")), 9294)
manhattan_plot("../data/ligation_read_start_counts", "Enterobacter_kobei_MSB1_1B__chromosome", "#CB4D42", 5.977266212427293, expression(paste(italic("E. kobei"), " chromosome")), 4837926)
manhattan_plot("../data/ligation_read_start_counts", "Enterobacter_kobei_MSB1_1B__plasmid_1", "#CB4D42", 4.221674997070769, expression(paste(italic("E. kobei"), " plasmid 1")), 136482)
manhattan_plot("../data/ligation_read_start_counts", "Enterobacter_kobei_MSB1_1B__plasmid_2", "#CB4D42", 4.267171728403014, expression(paste(italic("E. kobei"), " plasmid 2")), 108411)
manhattan_plot("../data/ligation_read_start_counts", "Enterobacter_kobei_MSB1_1B__plasmid_3", "#CB4D42", 2.8920946026904804, expression(paste(italic("E. kobei"), " plasmid 3")), 4665)
manhattan_plot("../data/ligation_read_start_counts", "Enterobacter_kobei_MSB1_1B__plasmid_4", "#CB4D42", 2.7781512503836434, expression(paste(italic("E. kobei"), " plasmid 4")), 3715)
manhattan_plot("../data/ligation_read_start_counts", "Haemophilus_unknown_M1C132_1__chromosome", "#CB4D42", 5.612889769287484, expression(paste(italic("Haemophilus"), " chromosome")), 2051882)
manhattan_plot("../data/ligation_read_start_counts", "Haemophilus_unknown_M1C132_1__plasmid_1", "#CB4D42", 3.895422546039408, expression(paste(italic("Haemophilus"), " plasmid 1")), 39398)
manhattan_plot("../data/ligation_read_start_counts", "Haemophilus_unknown_M1C132_1__plasmid_2", "#CB4D42", 3.330413773349191, expression(paste(italic("Haemophilus"), " plasmid 2")), 10719)
manhattan_plot("../data/ligation_read_start_counts", "Haemophilus_unknown_M1C132_1__plasmid_3", "#CB4D42", 2.9912260756924947, expression(paste(italic("Haemophilus"), " plasmid 3")), 9975)
manhattan_plot("../data/ligation_read_start_counts", "Klebsiella_oxytoca_MSB1_2C__chromosome", "#CB4D42", 6.064113412115037, expression(paste(italic("K. oxytoca"), " chromosome")), 5804453)
manhattan_plot("../data/ligation_read_start_counts", "Klebsiella_oxytoca_MSB1_2C__plasmid_1", "#CB4D42", 4.37254380075907, expression(paste(italic("K. oxytoca"), " plasmid 1")), 118161)
manhattan_plot("../data/ligation_read_start_counts", "Klebsiella_oxytoca_MSB1_2C__plasmid_2", "#CB4D42", 4.067442842776381, expression(paste(italic("K. oxytoca"), " plasmid 2")), 58472)
manhattan_plot("../data/ligation_read_start_counts", "Klebsiella_oxytoca_MSB1_2C__plasmid_3", "#CB4D42", 2.9542425094393248, expression(paste(italic("K. oxytoca"), " plasmid 3")), 4574)
manhattan_plot("../data/ligation_read_start_counts", "Klebsiella_variicola_INF345__chromosome", "#CB4D42", 6.033721225732489, expression(paste(italic("K. variicola"), " chromosome")), 5417255)
manhattan_plot("../data/ligation_read_start_counts", "Klebsiella_variicola_INF345__plasmid_1", "#CB4D42", 4.693726948923647, expression(paste(italic("K. variicola"), " plasmid 1")), 250980)
manhattan_plot("../data/ligation_read_start_counts", "Klebsiella_variicola_INF345__plasmid_2", "#CB4D42", 4.476251796007033, expression(paste(italic("K. variicola"), " plasmid 2")), 243620)
manhattan_plot("../data/ligation_read_start_counts", "Klebsiella_variicola_INF345__plasmid_3", "#CB4D42", 3.8020892578817325, expression(paste(italic("K. variicola"), " plasmid 3")), 31780)
manhattan_plot("../data/ligation_read_start_counts", "Klebsiella_variicola_INF345__plasmid_4", "#CB4D42", 3.0569048513364727, expression(paste(italic("K. variicola"), " plasmid 4")), 5783)
manhattan_plot("../data/ligation_read_start_counts", "Klebsiella_variicola_INF345__plasmid_5", "#CB4D42", 2.845098040014257, expression(paste(italic("K. variicola"), " plasmid 5")), 3514)
manhattan_plot("../data/ligation_read_start_counts", "Serratia_marcescens_17-147-1671__chromosome", "#CB4D42", 6.0354377435390045, expression(paste(italic("S. marcescens"), " chromosome")), 5518076)
manhattan_plot("../data/ligation_read_start_counts", "Serratia_marcescens_17-147-1671__plasmid_1", "#CB4D42", 4.392345155361204, expression(paste(italic("S. marcescens"), " plasmid 1")), 184477)
manhattan_plot("../data/ligation_read_start_counts", "Serratia_marcescens_17-147-1671__plasmid_2", "#CB4D42", 4.45969397647797, expression(paste(italic("S. marcescens"), " plasmid 2")), 161385)
manhattan_plot("../data/ligation_read_start_counts", "Serratia_marcescens_17-147-1671__plasmid_4", "#CB4D42", 2.57978359661681, expression(paste(italic("S. marcescens"), " plasmid 4")), 1934)

manhattan_plot("../data/rapid_read_start_counts", "Acinetobacter_baumannii_J9__chromosome", "#00989D", 5.879233546813747, expression(paste(italic("A. baumannii"), " chromosome")), 3798646)
manhattan_plot("../data/rapid_read_start_counts", "Acinetobacter_baumannii_J9__plasmid_1", "#00989D", 4.418301291319746, expression(paste(italic("A. baumannii"), " plasmid 1")), 145059)
manhattan_plot("../data/rapid_read_start_counts", "Acinetobacter_baumannii_J9__plasmid_2", "#00989D", 3.0791812460476247, expression(paste(italic("A. baumannii"), " plasmid 2")), 6078)
manhattan_plot("../data/rapid_read_start_counts", "Citrobacter_koseri_MINF_9D__chromosome", "#00989D", 5.976881629799957, expression(paste(italic("C. koseri"), " chromosome")), 4758908)
manhattan_plot("../data/rapid_read_start_counts", "Citrobacter_koseri_MINF_9D__plasmid_1", "#00989D", 4.083502619830267, expression(paste(italic("C. koseri"), " plasmid 1")), 64962)
manhattan_plot("../data/rapid_read_start_counts", "Citrobacter_koseri_MINF_9D__plasmid_2", "#00989D", 3.2648178230095364, expression(paste(italic("C. koseri"), " plasmid 2")), 9294)
manhattan_plot("../data/rapid_read_start_counts", "Enterobacter_kobei_MSB1_1B__chromosome", "#00989D", 5.977266212427293, expression(paste(italic("E. kobei"), " chromosome")), 4837926)
manhattan_plot("../data/rapid_read_start_counts", "Enterobacter_kobei_MSB1_1B__plasmid_1", "#00989D", 4.221674997070769, expression(paste(italic("E. kobei"), " plasmid 1")), 136482)
manhattan_plot("../data/rapid_read_start_counts", "Enterobacter_kobei_MSB1_1B__plasmid_2", "#00989D", 4.267171728403014, expression(paste(italic("E. kobei"), " plasmid 2")), 108411)
manhattan_plot("../data/rapid_read_start_counts", "Enterobacter_kobei_MSB1_1B__plasmid_3", "#00989D", 2.8920946026904804, expression(paste(italic("E. kobei"), " plasmid 3")), 4665)
manhattan_plot("../data/rapid_read_start_counts", "Enterobacter_kobei_MSB1_1B__plasmid_4", "#00989D", 2.7781512503836434, expression(paste(italic("E. kobei"), " plasmid 4")), 3715)
manhattan_plot("../data/rapid_read_start_counts", "Haemophilus_unknown_M1C132_1__chromosome", "#00989D", 5.612889769287484, expression(paste(italic("Haemophilus"), " chromosome")), 2051882)
manhattan_plot("../data/rapid_read_start_counts", "Haemophilus_unknown_M1C132_1__plasmid_1", "#00989D", 3.895422546039408, expression(paste(italic("Haemophilus"), " plasmid 1")), 39398)
manhattan_plot("../data/rapid_read_start_counts", "Haemophilus_unknown_M1C132_1__plasmid_2", "#00989D", 3.330413773349191, expression(paste(italic("Haemophilus"), " plasmid 2")), 10719)
manhattan_plot("../data/rapid_read_start_counts", "Haemophilus_unknown_M1C132_1__plasmid_3", "#00989D", 2.9912260756924947, expression(paste(italic("Haemophilus"), " plasmid 3")), 9975)
manhattan_plot("../data/rapid_read_start_counts", "Haemophilus_unknown_M1C132_1__plasmid_4", "#00989D", 3.164352855784437, expression(paste(italic("Haemophilus"), " plasmid 4")), 7392)
manhattan_plot("../data/rapid_read_start_counts", "Haemophilus_unknown_M1C132_1__plasmid_5", "#00989D", 3.0492180226701815, expression(paste(italic("Haemophilus"), " plasmid 5")), 5675)
manhattan_plot("../data/rapid_read_start_counts", "Klebsiella_oxytoca_MSB1_2C__chromosome", "#00989D", 6.064113412115037, expression(paste(italic("K. oxytoca"), " chromosome")), 5804453)
manhattan_plot("../data/rapid_read_start_counts", "Klebsiella_oxytoca_MSB1_2C__plasmid_1", "#00989D", 4.37254380075907, expression(paste(italic("K. oxytoca"), " plasmid 1")), 118161)
manhattan_plot("../data/rapid_read_start_counts", "Klebsiella_oxytoca_MSB1_2C__plasmid_2", "#00989D", 4.067442842776381, expression(paste(italic("K. oxytoca"), " plasmid 2")), 58472)
manhattan_plot("../data/rapid_read_start_counts", "Klebsiella_oxytoca_MSB1_2C__plasmid_3", "#00989D", 2.9542425094393248, expression(paste(italic("K. oxytoca"), " plasmid 3")), 4574)
manhattan_plot("../data/rapid_read_start_counts", "Klebsiella_variicola_INF345__chromosome", "#00989D", 6.033721225732489, expression(paste(italic("K. variicola"), " chromosome")), 5417255)
manhattan_plot("../data/rapid_read_start_counts", "Klebsiella_variicola_INF345__plasmid_1", "#00989D", 4.693726948923647, expression(paste(italic("K. variicola"), " plasmid 1")), 250980)
manhattan_plot("../data/rapid_read_start_counts", "Klebsiella_variicola_INF345__plasmid_2", "#00989D", 4.476251796007033, expression(paste(italic("K. variicola"), " plasmid 2")), 243620)
manhattan_plot("../data/rapid_read_start_counts", "Klebsiella_variicola_INF345__plasmid_3", "#00989D", 3.8020892578817325, expression(paste(italic("K. variicola"), " plasmid 3")), 31780)
manhattan_plot("../data/rapid_read_start_counts", "Klebsiella_variicola_INF345__plasmid_4", "#00989D", 3.0569048513364727, expression(paste(italic("K. variicola"), " plasmid 4")), 5783)
manhattan_plot("../data/rapid_read_start_counts", "Klebsiella_variicola_INF345__plasmid_5", "#00989D", 2.845098040014257, expression(paste(italic("K. variicola"), " plasmid 5")), 3514)
manhattan_plot("../data/rapid_read_start_counts", "Serratia_marcescens_17-147-1671__chromosome", "#00989D", 6.0354377435390045, expression(paste(italic("S. marcescens"), " chromosome")), 5518076)
manhattan_plot("../data/rapid_read_start_counts", "Serratia_marcescens_17-147-1671__plasmid_1", "#00989D", 4.392345155361204, expression(paste(italic("S. marcescens"), " plasmid 1")), 184477)
manhattan_plot("../data/rapid_read_start_counts", "Serratia_marcescens_17-147-1671__plasmid_2", "#00989D", 4.45969397647797, expression(paste(italic("S. marcescens"), " plasmid 2")), 161385)
manhattan_plot("../data/rapid_read_start_counts", "Serratia_marcescens_17-147-1671__plasmid_4", "#00989D", 2.57978359661681, expression(paste(italic("S. marcescens"), " plasmid 4")), 1934)
```
